File: C/Baekjoon/.1000/1761.cpp
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    int N, M, a, b, c, diff, j, x, y;
    cin >> N;

    vector<pair<int, int>> arr[N + 1];
    bool checked[N + 1] = {};

    for (int i = 0; i < N - 1; i++)
    {
        cin >> a >> b >> c;
        arr[a].push_back(make_pair(b, c));
        arr[b].push_back(make_pair(a, c));
    }

    stack<int> st;
    int parent[N + 1][17] = {}, depth[N + 1] = {}, length[N + 1] = {};

    st.push(1);
    checked[1] = true;

    while (!st.empty())
    {
        a = st.top();
        st.pop();

        for (pair<int, int> i : arr[a])
        {
            if (checked[i.first])
                continue;

            parent[i.first][0] = a;
            checked[i.first] = true;
            depth[i.first] = depth[a] + 1;
            length[i.first] = length[a] + i.second;

            for (int j = 1; 1 << j <= N; j++)
            {
                if (!parent[i.first][j - 1])
                    break;
                parent[i.first][j] = parent[parent[i.first][j - 1]][j - 1];
            }

            st.push(i.first);
        }
    }

    cin >> M;
    for (int i = 0; i < M; i++)
    {
        cin >> a >> b;
        if (depth[a] < depth[b])
        {
            x = b;
            y = a;
        }
        else
        {
            x = a;
            y = b;
        }

        diff = depth[x] - depth[y];
        j = 0;

        while (diff)
        {
            if (diff % 2)
                x = parent[x][j];
            j++;
            diff /= 2;
        }

        for (int j = 16; j >= 0; j--)
        {
            if (parent[x][j] != parent[y][j])
            {
                x = parent[x][j];
                y = parent[y][j];
            }
        }

        if (x != y)
            x = parent[x][0];

        cout << length[a] + length[b] - length[x] * 2 << '\n';
    }
}

File: PS/Baekjoon/.1000/1761.py
from sys import stdin
from collections import deque

N = int(stdin.readline())
arr = [[] for i in range(N + 1)]

for i in range(N - 1):
    A, B, C = map(int, stdin.readline().rstrip().split())
    arr[A].append((B, C))
    arr[B].append((A, C))

st = deque([1])
checked = [False for i in range(N + 1)]
checked[1] = True

depth = [0 for i in range(N + 1)]
length = [0 for i in range(N + 1)]
parent = [[0 for j in range(18)] for i in range(N + 1)]

while len(st):
    p = st.pop()
    for q, l in arr[p]:
        if checked[q]:
            continue

        checked[q] = True
        parent[q][0] = p
        depth[q] = depth[p] + 1
        length[q] = length[p] + l

        k = 1
        while 2**k <= N:
            if not parent[q][k - 1]:
                break
            parent[q][k] = parent[parent[q][k - 1]][k - 1]
            k += 1

        st.append(q)

M = int(stdin.readline())
for i in range(M):
    A, B = map(int, stdin.readline().rstrip().split())
    res = length[A] + length[B]
    if depth[B] > depth[A]:
        A, B = B, A

    diff = depth[A] - depth[B]
    for j in range(17, -1, -1):
        if 2**j <= diff:
            diff -= 2**j
            A = parent[A][j]

    for j in range(17, -1, -1):
        if parent[A][j] != parent[B][j]:
            A, B = parent[A][j], parent[B][j]
    if A != B:
        A = parent[A][0]

    print(res - length[A] * 2)
